src/gtk/toga_gtk/widgets/canvas.py

```diff
@@ -33,69 +28,51 @@ def create(self):
 
         self.native = Gtk.DrawingArea()
         self.native.interface = self.interface
-        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, self.native.get_allocated_width(),
-                                     self.native.get_allocated_height())
-        self.native.context = cairo.Context(surface)
+        self.surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, self.native.get_allocated_width(),
+                                          self.native.get_allocated_height())
+        self.native_context = cairo.Context(self.surface)
         self.native.font = None
 
-    def set_on_draw(self, handler):
-        self.native.connect('draw', handler)
-
-    def set_context(self, context):
-        self.native.context = context
+    def context(self, context):
+        def draw_callback(canvas, native_context):
+            self.native_context = native_context
```

Comment:
This bit seems weird - doesn't this mean the Impl Canvas will have it's native_context object overwriten every time a new context is created?